source/ceylon/future/Future.ceylon

```diff
@@ -0,0 +1,92 @@
+import ceylon.future.util {
+    Latch,
+    AcquireTimeoutException
+}
+import ceylon.promise {
+    Promise,
+    ExecutionContext
+}
+
+"Wraps a [[Promise]] that will be completed by another thread."
+see (`function submit`)
+shared class Future<Value>(delegate)
+        extends Promise<Value>() {
+    "The concurrently completed [[Promise]]."
+    Promise<Value> delegate;
+    
+    /*
+     * Delegation
+     */
+    
+    shared actual ExecutionContext context => delegate.context;
+    
+    shared actual Promise<Result> map<Result>(
+        Result onFulfilled(Value val),
+        Result onRejected(Throwable reason))
+            => delegate.map(onFulfilled, onRejected);
+    
+    shared actual Promise<Result> flatMap<Result>(
+        Promise<Result> onFulfilled(Value val),
+        Promise<Result> onRejected(Throwable reason))
+            => delegate.flatMap(onFulfilled, onRejected);
+    
+    /*
+     * New functionality
+     */
+    
+    value completion = Latch();
+    
+    variable Boolean _done = false;
```

Comment:
> you need synchronisation because the thread that updates this value in onFufilled/onRejected is not the same thread that will uses the Future::done.

Looking at the source of AtomicBoolean, it does just use volatile with no synchronisation on the `get` and `set`. Since Ceylon doesn't appear to have volatile yet, I will switch this to AtomicBoolean.